radix-engine/src/kernel/call_frame.rs

```diff
@@ -484,6 +536,150 @@ impl<C, L: Clone> CallFrame<C, L> {
+        for (_partition_number, module) in &node_substates {
+            for (substate_key, substate_value) in module {
+                let io_write = self
+                    .process_substate(destination_device, substate_value, None)
+                    .map_err(|e| CallbackError::Error(CreateNodeError::ProcessSubstateError(e)))?;
+
+                move_nodes_from_heap.extend(io_write.move_nodes_from_heap);
+                add_non_global_refs.extend(io_write.add_non_global_refs);
+
+                self.process_input_substate_key(substate_key).map_err(|e| {
+                    CallbackError::Error(CreateNodeError::ProcessSubstateKeyError(e))
+                })?;
+            }
+        }
+
+        match destination_device {
+            SubstateDevice::Store => {
+                self.stable_references
+                    .insert(node_id, StableReferenceType::Global);
+            }
+            SubstateDevice::Heap => {
+                self.owned_root_nodes.insert(node_id);
+            }
+        }
+
+        let mut handler = WrapperHandler {
+            call_frame: self,
+            handler,
+            phantom: PhantomData::default(),
+        };
+
+        substate_io.create_node(
+            &mut handler,
+            destination_device,
+            node_id,
+            node_substates,
+            SubstateIOWrite {
+                move_nodes_from_heap,
+                add_non_global_refs,
+                remove_non_global_refs: index_set_new(),
+            },
+        )?;
+
+        Ok(())
+    }
+
+    /// Removes node from call frame and owned nodes will be possessed by this call frame.
+    pub fn drop_node<S: SubstateStore>(
+        &mut self,
+        substate_io: &mut SubstateIO<S>,
+        node_id: &NodeId,
+    ) -> Result<NodeSubstates, DropNodeError> {
+        self.take_node_internal(node_id)
+            .map_err(DropNodeError::TakeNodeError)?;
+
+        let node_substates = substate_io.drop_node(SubstateDevice::Heap, node_id)?;
+
+        for (_, module) in &node_substates {
+            for (_, substate_value) in module {
+                //=============
+                // Process own
+                //=============
+                for own in substate_value.owned_nodes() {
+                    self.owned_root_nodes.insert(own.clone());
+                }
+
+                //====================
+                // Process references
+                //====================
+                for reference in substate_value.references() {
+                    if reference.is_global() {
+                        // Expand stable references
+                        // We keep all global references even if the owning substates are dropped.
+                        // Revisit this if the reference model is changed.
+                        self.stable_references
+                            .insert(reference.clone(), StableReferenceType::Global);
+                    }
```

Comment:
`transient_references` should not be existent for this node if it's droppable (otherwise there would be opened substates on this node and the `substate_io.drop_node()` call will fail.